### PR TITLE
Porting of Pressure Equalizer feature from Prusa Slicer 2.6.x

### PR DESCRIPTION
**Pull request by igiannakas**
This PR contains a port plus extension of the pressure equaliser feature from Prusa slicer. It also contains the fixes made to pressure equalizer by [MGunlogson](https://github.com/MGunlogson) in the below PR https://github.com/prusa3d/PrusaSlicer/pull/9622

**In summary:** 
This feature smooths out sudden extrusion rate changes that happen when the printer transitions from printing a high flow (high speed/larger width) extrusion to a lower flow (lower speed/smaller width) extrusion and vice versa. This feature works complementary to pressure advance/linear advance and is particularly useful in improving surface finish for:
1. Seams
2. Transition to overhangs
3. Transition to overhangs slowed down even further by the curled edges algorithm
4. Transitioning from a fast internal to a slower external perimeter

To support the above feature a new configuration variable has been added to the UI, under the Speed -> Advanced section called **Extrusion Rate Smoothing**. It is set to 0 by default, so the user actively elects to use this feature in prints that may benefit from it the most.

The above variable defines the maximum rate by which the extruded volumetric flow in mm3/sec can change over time. Higher values mean higher extrusion rate changes are allowed, resulting in faster speed transitions. A value of 0 disables the feature.

This parameter disables arc fitting as arcs are by default large segments of continuous speed and this feature needs them broken down to small segments in order to adjust the speed. When a value is entered in the above extrusion rate smoothing, arc fitting is automatically disabled and greyed out. When a 0 value is entered (i.e. the feature is disabled) arc fitting becomes available again.

**Before:**
Screenshot 2023-09-18 at 22 44 15

**After:**
Screenshot 2023-09-18 at 22 44 26

IMG_2436

@SoftFever I'd appreciate your review and thoughts.

Also, it would be fantastic if you could get a one off build in the pipeline for the pretty active team at issue #44 to test, if possible?

### CODE REVIEW

**Comment by SoftFever** [comment]
Well done!
Build is kicked off.

**Comment by igiannakas** [comment]
Thank you for this!

Please don’t merge it yet - I want to do some testing with my Bambu printer and maybe some parameter tuning but am away for the rest of the week and won’t get to it till the weekend.

**Comment by igiannakas** [comment]
Additional parameter added to the GUI that influences the granularity of the flow adjustment. A lower segment value results in a more granular speed adjustment to achieve the desired flow rate smoothing curve. However it also increases the size of the gcode file and the volume of gcode commands emitted. 

The default value of 2 should work well for most printers; but this value may need increasing for slower printers or if the command buffer cannot keep up on a bambu lab. 

**Value of 1 - gradual speed adjustments**
Screenshot 2023-09-19 at 18 52 06

**Value of 3 - less steps in the adjustment**
Screenshot 2023-09-19 at 18 52 19

**Comment by SoftFever** [comment]
Thank you!
I'm back and recovered from busy week-long event.
Will review and test soon 👍

**Comment by igiannakas** [comment]
> Thank you! I'm back and recovered from busy week-long event. Will review and test soon 👍

Same here, back from a week off, I’ve tested one complex model, worked well. Will be doing some more testing tomorrow.

**Comment by igiannakas** [comment]
So I've done some testing on the above. Algorithm works well with my X1 carbon, even with arc fitting disabled. Printer is a bit louder but for certain models the increase in quality due to the smoother ramp up in speeds is worth it.

Tweaked the tool tips according to my latest tests. In summary this is a useful feature for Bambu in certain edge cases - where there are significant slowdowns in the external perimeter that are present on a smooth surface, like the hull of the benchy.

Extrusion rate slope tests below:

**1. Disabled** - can clearly see where the Bambu pressure advance is needing some help to reduce the extrusion inconsistency when rapidly slowing down for an overhang
ERS Disabled

**2. Enabled - 250mm3/s2 -** Smoothed out completely. However can do better as the print is slowed down quite a bit in areas where it doesnt appear to be as necessary
ERS 250

**3. Enabled - 350mm3/s2 -** Limited/no artefacts and the minimum print speed reduction. Some minor artefacts re-appear but nothing too noticeable.
ERS 350

**4. Enabled - 500mm3/s2 -** Artefacts reappearing and becoming more noticeable, so need to back down slightly. 
ERS 500

So in summary, for the X1C, my recommended setting would be ~300-350mm3/s2 but only in prints that may benefit from it, i.e. ones that have significant variations in speed for the external perimeters.

Tool tips and default values updated accordingly to reflect the above

**Comment by SoftFever** [comment]
@igiannakas 
I have some questions about the changes to the parameters. The parameters used in PrusaSlicer are straightforward for me; they describe how quickly the volumetric speed can change. In other words, it's the derivative of the volumetric flow rate, hence the unit mm^3/s^2. When I set it to 3, and my layer height is 0.2 mm with a line width of 0.5 mm, I can expect that the Pressure Equalizer will take effect if the speed change is more than (3/(0.2*0.5)) = 30 mm/s within one second.

Back to this PR, I'm a bit confused about the parameter we used. ` "5x to 10x the maximum volumetric flow rate the hot end is capable off under ideal conditions. This allows "
    				 "for just enough smoothing to assist pressure advance in areas where sudden flow changes happen. A value of "
    				 "250-300 is a good starting point.\n\n"` 

When I set to 5, it gave me this result:
image

**Comment by igiannakas** [comment]
> @igiannakas I have some questions about the changes to the parameters. The parameters used in PrusaSlicer are straightforward for me; they describe how quickly the volumetric speed can change. In other words, it's the derivative of the volumetric flow rate, hence the unit mm^3/s^2. When I set it to 3, and my layer height is 0.2 mm with a line width of 0.5 mm, I can expect that the Pressure Equalizer will take effect if the speed change is more than (3/(0.2*0.5)) = 30 mm/s within one second.
> 
> Back to this PR, I'm a bit confused about the parameter we used. ` "5x to 10x the maximum volumetric flow rate the hot end is capable off under ideal conditions. This allows " "for just enough smoothing to assist pressure advance in areas where sudden flow changes happen. A value of " "250-300 is a good starting point.\n\n"`
> 
> When I set to 5, it gave me this result: image

I may have confused you - the parameters (annd behaviour) are identical to prusa, the tool tip was different. I’ve removed the reference to the volumetric flow rate in the latest tool tip.

**Comment by SoftFever** [comment]
The behavior is still a bit strange. (see screenshot in my previous commend, project attached here)
[PE.zip](https://github.com/SoftFever/OrcaSlicer/files/12708855/PE.zip)

Also the number in the tooltip also doesn't sound's right me too. It's way too big

**Comment by igiannakas** [comment]
> The behavior is still a bit strange. (see screenshot in my previous commend, project attached here) [PE.zip](https://github.com/SoftFever/OrcaSlicer/files/12708855/PE.zip)
> 
> Also the number in the tooltip also doesn't sound's right me too. It's way too big

Let me try and explain and please do correct me if I'm wrong :)  

The code looks for flow rate changes and smooths them out so that the change in flow rate over time is limited to a ceiling set by the slope. 

Say your printer is going all out, printing at 30mm3/sec volumetric speed, which is around 350mm/sec speed on a 0.2 layer height and 0.42 layer width.

Say you slow down to 20mm/sec speed for the tip of the benchy. This would slow you down to 1.68mm3/sec flow.

So you need (30-1.68)/(slope negative time) to decelerate to that volumetric speed -> 0.09 seconds or so at 320 slope negative. We can assume that the speed will go down linearly, hence your average speed over that deceleration is going to be (350-20)/2 = 165mm/sec.

Hence you’ll need around 165*0.09= 14.85mm distance to slow down (very roughly). 





For a slope of 3, then it would take 9.4 seconds, hence around 1.5meters to slow down, hence you’ll never reach max speed.

**Comment by SoftFever** [comment]
> > The behavior is still a bit strange. (see screenshot in my previous commend, project attached here) [PE.zip](https://github.com/SoftFever/OrcaSlicer/files/12708855/PE.zip)
> > Also the number in the tooltip also doesn't sound's right me too. It's way too big
> 
> Let me try and explain and please do correct me if I'm wrong :)
> 
> The code looks for flow rate changes and smooths them out so that the change in flow rate over time is limited to a ceiling set by the slope.
> 
> Say your printer is going all out, printing at 30mm3/sec volumetric speed, which is around 350mm/sec speed on a 0.2 layer height and 0.42 layer width.
> 
> Say you slow down to 20mm/sec speed for the tip of the benchy. This would slow you down to 1.68mm3/sec flow.
> 
> So you need (30-1.68)/(slope negative time) to decelerate to that volumetric speed -> 0.09 seconds or so at 320 slope negative. We can assume that the speed will go down linearly, hence your average speed over that deceleration is going to be (350-20)/2 = 165mm/sec.
> 
> Hence you’ll need around 165*0.09= 14.85mm distance to slow down (very roughly).
> 
> 
> 
> For a slope of 3, then it would take 9.4 seconds, hence around 1.5meters to slow down, hence you’ll never reach max speed.
> 
> The slow downs you see in the model are due to the overhangs and bridges - the printer slowing down in advance of these at a rate of slow down of X mm3/sec per second.

We are on same page.
I took a closer look just now, I found the latest PrusaSlicer explictly exclude external perameter from adjusting flow rate.
Hence the different behavior.
If you don't mind I can push the minor changes with latest PS changes to this branch.

**Comment by igiannakas** [comment]
> > > The behavior is still a bit strange. (see screenshot in my previous commend, project attached here) [PE.zip](https://github.com/SoftFever/OrcaSlicer/files/12708855/PE.zip)
> > > Also the number in the tooltip also doesn't sound's right me too. It's way too big
> > 
> > 
> > Let me try and explain and please do correct me if I'm wrong :)
> > The code looks for flow rate changes and smooths them out so that the change in flow rate over time is limited to a ceiling set by the slope.
> > Say your printer is going all out, printing at 30mm3/sec volumetric speed, which is around 350mm/sec speed on a 0.2 layer height and 0.42 layer width.
> > Say you slow down to 20mm/sec speed for the tip of the benchy. This would slow you down to 1.68mm3/sec flow.
> > So you need (30-1.68)/(slope negative time) to decelerate to that volumetric speed -> 0.09 seconds or so at 320 slope negative. We can assume that the speed will go down linearly, hence your average speed over that deceleration is going to be (350-20)/2 = 165mm/sec.
> > Hence you’ll need around 165*0.09= 14.85mm distance to slow down (very roughly).
> > 
> > For a slope of 3, then it would take 9.4 seconds, hence around 1.5meters to slow down, hence you’ll never reach max speed.
> > The slow downs you see in the model are due to the overhangs and bridges - the printer slowing down in advance of these at a rate of slow down of X mm3/sec per second.
> 
> We are on same page. I took a closer look just now, I found the latest PrusaSlicer explictly exclude external perameter from adjusting flow rate. Hence the different behavior. If you don't mind I can push the minor changes with latest PS changes to this branch.

Sure thing, push to the branch I’ll test them too.

thank you very much!

**Comment by igiannakas** [comment]
Ps. The core pressure equalizer files (cpp & hpp) were from 2.6.1 extended to include external perimeters + the fixes from https://github.com/MGunlogson in his PR to prusa.

File: src/libslic3r/GCode/PressureEqualizer.cpp [hunk 1]
```diff
@@ -619,8 +558,7 @@ void PressureEqualizer::adjust_volumetric_rate(const size_t fist_line_idx, const
                 // Limit by the succeeding volumetric flow rate.
                 rate_end = rate_succ;
 
-            // don't alter the flow rate for these extrusion types
-            if (!line.adjustable_flow || line.extrusion_role == GCodeExtrusionRole::BridgeInfill || line.extrusion_role == GCodeExtrusionRole::Ironing) {
+            if (!line.adjustable_flow || line.extrusion_role == GCodeExtrusionRole::ExternalPerimeter || line.extrusion_role == GCodeExtrusionRole::GapFill || line.extrusion_role == GCodeExtrusionRole::BridgeInfill || line.extrusion_role == GCodeExtrusionRole::Ironing) {
```
Comment (SoftFever) [inline src/libslic3r/GCode/PressureEqualizer.cpp:561]: @igiannakas 
One of the difference is that PS won't apply Pressure Eq to ExternalPerimeter and GapFill either.
GapFill is understandable.

I'm not too sure about why don't apply to ExternalPerimeter thought.
You have done the extensive tests on your side, I'll leave it to you to make the judgement call whether to execlude external perimeters.

**Comment by SoftFever** [comment]
> Ps. The core pressure equalizer files (cpp & hpp) were from 2.6.1 extended to include external perimeters + the fixes from https://github.com/MGunlogson in his PR to prusa.

Ah, ICIC
Seems the difference is coming from his changes.
Feel free to revert my latest commit then

File: src/libslic3r/GCode/PressureEqualizer.cpp (hunk 1 above)
Comment (igiannakas) [reply src/libslic3r/GCode/PressureEqualizer.cpp:561]: I’d think it’s best to leave the external perimeter in the code, it’s where the most benefit is :) - let me test without the changes from the above user but with external perimeter enabled and I’ll revert back.

**Comment by igiannakas** [comment]
Hi @SoftFever 

I'd recommend that the changes are reverted back the commit [131fb94](https://github.com/SoftFever/OrcaSlicer/pull/2161/commits/131fb94c6bebe0a6abb3ca28d4a162aacbe75f40) to the previous submitted version of the code. Reasons below:

1. Prusa default code does not do any pressure equalisation on external perimeters. This is where most artefacts are present so that is a miss on Prusa's side
2. Prusa continues equalising pressure even after a long travel move between two extrusion moves. This is not right as nozzle pressure drops to zero after a long travel move, especially after a retraction command has been issued. 

The changes from https://github.com/prusa3d/PrusaSlicer/pull/9622 that have been applied to this code address the above shortcomings. So I'd recommend we revert back to the previous commit until Prusa elects to do something about the above two...

Thoughts?

**Comment by SoftFever** [comment]
> Hi @SoftFever
> 
> I'd recommend that the changes are reverted back the commit [131fb94](https://github.com/SoftFever/OrcaSlicer/pull/2161/commits/131fb94c6bebe0a6abb3ca28d4a162aacbe75f40) to the previous submitted version of the code. Reasons below:
> 
> 1. Prusa default code does not do any pressure equalisation on external perimeters. This is where most artefacts are present so that is a miss on Prusa's side
> 2. Prusa continues equalising pressure even after a long travel move between two extrusion moves. This is not right as nozzle pressure drops to zero after a long travel move, especially after a retraction command has been issued.
> 
> The changes from [prusa3d/PrusaSlicer#9622](https://github.com/prusa3d/PrusaSlicer/pull/9622) that have been applied to this code address the above shortcomings. So I have reverted back to the previous commit until Prusa elects to do something about the above two issues...
> 
> Thoughts?

Make perfect sense to me

**Comment by igiannakas** [comment]
Awesome! So is it good to go or any changes needed? 🙏🏻

**Review by SoftFever — APPROVED** [review]
LGTM

**Comment by SoftFever** [comment]
> Awesome! So is it good to go or any changes needed? 🙏🏻

Looks good now.
Gonna merge it now.
Thank you so much for porting this feature 👍

**Comment by igiannakas** [comment]
Your welcome! 🙏🏻
